### PR TITLE
Add interface for unregistering query

### PR DESCRIPTION
**Pull request by yufeiyu**

### CODE REVIEW

File: pkg/prediction/config/broadcast.go [hunk 1]
```diff
@@ -1,37 +1,41 @@
 package config
 
-type broadcast struct {
-	c chan broadcast
-	v interface{}
+type event struct {
```
Comment (qmhu) [inline pkg/prediction/config/broadcast.go:3]: Would you please add some unit test for it?

File: pkg/prediction/percentile/prediction.go [hunk 2]
```diff
@@ -2,22 +2,23 @@ package percentile
...
 )
 
 var _ prediction.Interface = &percentilePrediction{}
 
 type percentilePrediction struct {
 	prediction.GenericPrediction
-	a aggregateSignalMap
-	//mr config.Receiver
-	qr config.Receiver
+	a         map[string]*aggregateSignalMap
```
Comment (qmhu) [inline pkg/prediction/percentile/prediction.go:18]: s/a/signalMap/g

File: pkg/prediction/dsp/prediction.go [hunk 3]
```diff
@@ -27,17 +27,21 @@ const (
 
 type periodicSignalPrediction struct {
 	prediction.GenericPrediction
-	a           aggregateSignalMap
-	qr          config.Receiver
+	a           map[string] /*query expression*/ *aggregateSignalMap
```
Comment (qmhu) [inline pkg/prediction/dsp/prediction.go:30]: s/a/signalMap/g

File: pkg/prediction/generic.go [hunk 4]
```diff
@@ -57,27 +58,43 @@ func (p *GenericPrediction) WithProviders(providers map[string]providers.Interfa
...
-func AggregateSignalKey(id string, labels []common.Label) string {
+func (p *GenericPrediction) DeleteQuery(queryExpr string) error {
+	if queryExpr == "" {
+		return fmt.Errorf("empty query expression")
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if _, exists := p.querySet[queryExpr]; exists {
```
Comment (qmhu) [inline pkg/prediction/generic.go:85]: Consider this situation:
EHPA A is running for a pod named "test1", and meanwhile Recommendation B is recommend hpa for the same pod "test1", their queryExpr for cpu usage is same. if delete the query here, then should effect the EHPA ?

File: pkg/prediction/config/broadcast.go (hunk 1 above)
Comment (yufeiyu) [reply pkg/prediction/config/broadcast.go:3]: broadcast_test.go

File: pkg/prediction/dsp/prediction.go [hunk 5]
```diff
@@ -145,28 +149,56 @@ func (p *periodicSignalPrediction) Run(stopCh <-chan struct{}) {
...
-				if err := p.updateAggregateSignalsWithQuery(queryExpr); err != nil {
-					klog.ErrorS(err, "Failed to updateAggregateSignalsWithQuery.")
+	go func() {
+		for {
+			// Waiting for a WithQuery request
+			queryExpr := <-p.withCh
+			if _, ok := p.stopChMap.Load(queryExpr); ok {
+				continue
+			}
+			klog.InfoS("Register a query expression for prediction.", "queryExpr", queryExpr)
```
Comment (mfanjie) [inline pkg/prediction/dsp/prediction.go:159]: one common comment, add log level for debug info

File: pkg/prediction/generic.go (hunk 4 above)
Comment (yufeiyu) [reply pkg/prediction/generic.go]: Add caller to WithQuery/DeleteQuery

File: pkg/prediction/dsp/prediction.go (hunk 5 above)
Comment (yufeiyu) [reply pkg/prediction/dsp/prediction.go]: Added

File: pkg/prediction/dsp/aggregate_signal.go [hunk 6]
```diff
@@ -1,6 +1,8 @@
 package dsp
 
 import (
+	"github.com/gocrane/crane/pkg/prediction"
```
Comment (kitianFresh) [inline pkg/prediction/dsp/aggregate_signal.go]: import pkg

Comment (qmhu) [inline pkg/prediction/dsp/aggregate_signal.go:4]: import format

File: pkg/prediction/dsp/aggregate_signal.go [hunk 7]
```diff
@@ -26,3 +28,93 @@ func (a *aggregateSignal) setPredictedTimeSeries(ts *common.TimeSeries) {
 		a.lastUpdateTime = time.Now()
 	}
 }
+
+type aggregateSignals struct {
```
Comment (qmhu) [inline pkg/prediction/dsp/aggregate_signal.go:32]: Would you please add ut for this struct

File: pkg/prediction/dsp/prediction.go [hunk 8]
```diff
@@ -27,17 +27,21 @@ const (
 
 type periodicSignalPrediction struct {
 	prediction.GenericPrediction
-	a           aggregateSignalMap
-	qr          config.Receiver
+	a           aggregateSignals
```
Comment (qmhu) [inline pkg/prediction/dsp/prediction.go:30]: s/a/aggregateSignals/g

File: pkg/prediction/percentile/aggregate_signal.go [hunk 9]
```diff
@@ -1,7 +1,9 @@
 package percentile
 
 import (
+	"github.com/gocrane/crane/pkg/prediction"
```
Comment (qmhu) [inline pkg/prediction/percentile/aggregate_signal.go:4]: import format

File: pkg/prediction/percentile/aggregate_signal.go [hunk 10]
```diff
@@ -34,3 +36,93 @@ func newAggregateSignal(c *internalConfig) *aggregateSignal {
 		creationTime:    time.Now(),
 	}
 }
+
+type aggregateSignals struct {
```
Comment (qmhu) [inline pkg/prediction/percentile/aggregate_signal.go:40]: Please add ut for this struct

File: pkg/prediction/percentile/prediction.go [hunk 11]
```diff
@@ -2,22 +2,23 @@ package percentile
...
 )
 
 var _ prediction.Interface = &percentilePrediction{}
 
 type percentilePrediction struct {
 	prediction.GenericPrediction
-	a aggregateSignalMap
-	//mr config.Receiver
-	qr config.Receiver
+	a         aggregateSignals
```
Comment (qmhu) [inline pkg/prediction/percentile/prediction.go:18]: s/a/aggregateSignals/g

File: pkg/prediction/percentile/aggregate_signal.go (hunk 10 above)
Comment (yufeiyu) [reply pkg/prediction/percentile/aggregate_signal.go:40]: done

File: pkg/prediction/percentile/aggregate_signal.go (hunk 9 above)
Comment (yufeiyu) [reply pkg/prediction/percentile/aggregate_signal.go]: fixed

**Pull request merged**